firebase-installations/src/main/java/com/google/firebase/installations/local/IidStore.java

```diff
@@ -38,17 +40,79 @@
   private static final String IID_SHARED_PREFS_NAME = "com.google.android.gms.appid";
   private static final String STORE_KEY_PUB = "|S||P|";
   private static final String STORE_KEY_ID = "|S|id";
+  private static final String STORE_KEY_TOKEN = "|T|";
+  private static final String DEFAULT_SCOPE = "|*";
+  private static final String JSON_TOKEN_KEY = "token";
+  private static final String JSON_ENCODED_PREFIX = "{";
 
   @GuardedBy("iidPrefs")
   private final SharedPreferences iidPrefs;
 
+  private final String defaultSenderId;
+
   public IidStore() {
     // Different FirebaseApp in the same Android application should have the same application
     // context and same dir path. We only read existing Iids for the default firebase application.
     iidPrefs =
         FirebaseApp.getInstance()
             .getApplicationContext()
             .getSharedPreferences(IID_SHARED_PREFS_NAME, Context.MODE_PRIVATE);
+
+    defaultSenderId = getDefaultSenderId(FirebaseApp.getInstance());
+  }
+
+  private static String getDefaultSenderId(FirebaseApp app) {
+    // Check for an explicit sender id
+    String senderId = app.getOptions().getGcmSenderId();
+    if (senderId != null) {
+      return senderId;
+    }
+    String appId = app.getOptions().getApplicationId();
+    if (!appId.startsWith("1:")) {
+      // Not v1, server should be updated to accept the full app ID now
+      return appId;
+    }
+    // For v1 app IDs, fall back to parsing the project number out
+    @SuppressWarnings("StringSplitter")
+    String[] parts = appId.split(":");
+    if (parts.length < 2) {
```

Comment:
if method #split, splits all parts, then the result should contain exactly 4 parts...

if (parts.length != 4)

Reply:
Done.